### PR TITLE
[WIP] use base64 over base64-bytestring

### PR DESCRIPTION
**Pull request by chessai**
I added SIMD to base64 [(See PR)](https://github.com/emilypi/base64/pull/50).

Per profiling, chainweb-node spends about 6% of its time performing base64 decode on blocks. According to the microbenchmarks, this should hopefully offer a speedup of ~15x on decode of things that are the average block size.

The other nice thing about the `base64` library is that it has the option of being more type-safe.
All these assertions I've added here are unsafe, but, they were just implicit before.

### CODE REVIEW

**Comment by jwiegley** [comment]
Does this pass replay?

**Comment by chessai** [comment]
> Does this pass replay?

No; I think that we actually do need to keep `base64-bytestring` for certain operations where we care about the on-chain error messages. Right now, a replay fails immediately because of one such instance.

File: chainweb.cabal
Comment (larskuhtz) [inline chainweb.cabal:329]: 👍 the upper bound is tracked in the cabal.project file.

File: chainweb.cabal
Comment (larskuhtz) [inline chainweb.cabal:328]: Is this known to break with base64 >=0.6? Otherwise please remove, since we don't manually track "soft" upper bounds.

```suggestion
        , base64 >= 0.5
```

File: src/Chainweb/Mempool/Mempool.hs
Comment (larskuhtz) [inline src/Chainweb/Mempool/Mempool.hs:753]: I am a bit concerned about `B64.UrlPadded`. Does this mean the it uses padding? All use of base64 throughout chainweb should be base64Url without padding.

File: src/Chainweb/Utils.hs
Comment (larskuhtz) [inline src/Chainweb/Utils.hs:611]: Are the functions for the unpadded (and non-url) variants used anywhere in the code? If not we should consider removing them, so that nobody uses them. This is to keep used encodings simple and consistent.

**Review by larskuhtz — CHANGES_REQUESTED** [review]
We should double check that all use of base64 is the URL variant without padding.

Regarding replay: this should be fine since chainweb doesn't put base64 encoded things on-chain. Base64 is used only as textual representation of binary data (mostly in JSON). Merkle hashes of on-chain data are computed on the plain binary representation.

However, I am not 100% certain about the use of base64 within pact service. In theory it shouldn't affect on-chain data (in particular errors in pact service shouldn't be included on-chain), but we should double check that.

File: chainweb.cabal
Comment (larskuhtz) [inline chainweb.cabal:328]: It seems that version 0.5 of base64 hasn't yet been released. The latest version on Hackage is 0.4.2.4. 0.5 is only available as source repository from Github. If 0.5 is actually required, we would have to ask the maintainer (@emilypi?) to release 0.5 on Hackage or include it as source repository package in cabal.project.

Comment (emilypi) [reply chainweb.cabal:328]: It hasn't been released and I"m still working on the PR, which will significantly affect this PR.

**Comment by emilypi** [comment]
I like your enthusiasm @chessai! However, this is jumping the gun alittle bit. The PR that is going in to serve as the basis for `base64-0.5` is going to streamline alot of the functionality changed in this PR that does alot of redundant work (adding pads, removing pads, converting to/from text etc), so I ask that we hold off on it here until I'm confident I can nail down the api. 

That said, with the addition of SIMD support, I do endorse this change, and I would like to revert from `base64-bytestring` at some point, because it is _very_ suboptimal (by design, per Bos + HVR's request). 

> Does this pass replay?

It should @jwiegley, since we're going in at the chainweb level, and the problems with replay w.r.t. Base64 are with the Pact natives wholesale committing to the error format returned by `base64-bytestring`. I'm working on reverting those changes in the latest `base64` major version so that we can switch to that in Pact, but the changes here do not affect Pact, so they shouldn't affect block hashes.

**Comment by chessai** [comment]
> I like your enthusiasm @chessai! However, this is jumping the gun alittle bit. The PR that is going in to serve as the basis for `base64-0.5` is going to streamline alot of the functionality changed in this PR that does alot of redundant work (adding pads, removing pads, converting to/from text etc), so I ask that we hold off on it here until I'm confident I can nail down the api.
> 
> That said, with the addition of SIMD support, I do endorse this change, and I would like to revert from `base64-bytestring` at some point, because it is _very_ suboptimal (by design, per Bos + HVR's request).
> 
> > Does this pass replay?
> 
> It should @jwiegley, since we're going in at the chainweb level, and the problems with replay w.r.t. Base64 are with the Pact natives wholesale committing to the error format returned by `base64-bytestring`. I'm working on reverting those changes in the latest `base64` major version so that we can switch to that in Pact, but the changes here do not affect Pact, so they shouldn't affect block hashes.

Thanks,

Yeah, after talking to you yesterday/digging around in your PR/the codebase more, it seems like this is indeed jumping the gun.
I just saw those speedups and I was really excited.

I'll leave this open for now, because I do think it's something we should return to.

**Pull request closed without merging**
